### DISCUSSION TITLE
请问ws和tcp的host跟path伪装底层原理是什么？

### DISCUSSION THREAD

**Discussion by XY-ShadowK** [v2fly/v2ray-core#3221] [Q&A]
如题，网上只能找到配置教程，我另外的程序需要host跟path伪装功能，求大佬们尽量讲详细点，或者发下能在哪里找到详细原理。
感激不尽！

**Comment by tomandjerry27** [comment]
大体上有两种思路。
思路一：加密host然后把他隐藏起来。
比如vmess-tcp，数据段中是tcp包头，打开tcp是vmess协议头，打开vmess协议头是加密的数据。你问host和path呢？那当然是在被加密的数据里。
还比如ECH-tcp，也把全部数据加密了。
虽然这样墙就瞎了眼，但如果墙宁可错杀一千也不放过一个，那思路一还是会行不通的。
思路二就是你提到的伪装。但你提到的websocket是不加密的，所以我猜想你真正想说的大概是websocket-tls-tcp。
如果你不用科技，直接访问github。你用个抓包软件抓包这个访问过程。能看到一个dns请求，一个dns响应。三次tcp握手。然后是一个tls请求，host就明文（无加密）写在这个tls中。。。。。。（后续的通信过程略）如果你不是去github而是去pronhub呢？结果可想而知。
上面这个过程我想你应该已经知道，但为了方便比较用科技前后的变化，我还是多余说明一遍。
再讲websocket-tls-tcp。其中的websocket协议肚子里边被塞进了上面提到的那个写了明文github的tls数据。websocket又被塞进了我们准备的tls协议（对，还是他，tls。但我们这次需要准备好一个不论真假的域名，然后给这个域名准备一个证书。有了域名和证书，我们就能使用tls协议了。）中，tls协议给数据（即websocket）加了密，而明文host是我们准备的域名（这个域名会是个被墙放行且在海外的域名，比如apple。）。然后tls被塞进tcp协议。。。。。。（后续略）
这么一对比，不难发现思路二的核心点就在于把github给替换成了apple，这样便达成了伪装。
但如果墙宁可错杀一千也不放过一个，那思路二也是会行不通的。

**Reply by XY-ShadowK** [reply]
也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。

**Reply by tomandjerry27** [reply]
> 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。

你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。
但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。
另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？

**Reply by XY-ShadowK** [reply]
> > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> 
> 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？

自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。
主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？

**Reply by tomandjerry27** [reply]
> > > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> > 
> > 
> > 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？
> 
> 自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。 主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？

ws和tcp里都不放host的。host明文放在tls协议头中。path我不关心，毕竟host都明文了，path明不明根本不重要。
所以我一开始猜想你说的是ws-tls-tcp，没有tls的话，你只是能马赛克掉host，不能拿另一个host替换。那就变成思路一了。

**Reply by XY-ShadowK** [reply]
> > > > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> > > 
> > > 
> > > 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？
> > 
> > 
> > 自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。 主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？
> 
> ws和tcp里都不放host的。host明文放在tls协议头中。path我不关心，毕竟host都明文了，path明不明根本不重要。 所以我一开始猜想你说的是ws-tls-tcp，没有tls的话，你只是能马赛克掉host，不能拿另一个host替换。那就变成思路一了。

我查了tls不管是握手还是记录的报文头部都没有标放host的位置啊，ws跟host沾边的也只找到http报文里的host请求头。而且v2ray本身不开tls也是能伪装host跟path的，应该不是只有tls能用吧。
我又看了下v2ray的tcp传输时开伪装应该是用http报文伪装，可能是用http的host请求头？但是伪装类型也可以是none，实在搞不清楚了。

**Reply by tomandjerry27** [reply]
> > > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> > 
> > 
> > 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？
> 
> 自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。 主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？

path我想他是被加密后放在tls协议当中的。host是有特殊的缘由才会被特殊对待的（指明文放在tls协议头）。
你知道cdn吗？cdn就是让各地的人们从就近的服务器上交互数据，以此加快访问。
假设我在北京放一台cdn服务器，南京放一台cdn服务器。大陆北方网友刷抖音访问北京cdn，大陆南方网友刷抖音访问南京cdn。这就是一个cdn被应用的简单例子了。
这时候聪明的人就会想了，那一台cdn只能服务于抖音吗？能不能让他服务于更多app呢？正好抖音和今日头条都是字节跳动旗下的app（都是一家的），那就让大陆北方网友刷今日头条也去北京cdn，南方......去南京cdn。
这时候不难发现，抖音和今日头条域名不同，但北方/南方的网友数据都会去同一个IP地址（北京/南京cdn的IP）。
那么，对于cdn来说，怎么区分他得到的数据包呢？数据包该转交给抖音还是今日头条的服务器来处理呢？
于是，为了方便我们如今大规模地使用cdn。域名（host）被明文放在了tls协议头，说白了，这就是给cdn看的。
反观path呢？他有什么必要明文放在tls协议头?

**Reply by tomandjerry27** [reply]
> > > > > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> > > > 
> > > > 
> > > > 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？
> > > 
> > > 
> > > 自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。 主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？
> > 
> > 
> > ws和tcp里都不放host的。host明文放在tls协议头中。path我不关心，毕竟host都明文了，path明不明根本不重要。 所以我一开始猜想你说的是ws-tls-tcp，没有tls的话，你只是能马赛克掉host，不能拿另一个host替换。那就变成思路一了。
> 
> 我查了tls不管是握手还是记录的报文头部都没有标放host的位置啊，ws跟host沾边的也只找到http报文里的host请求头。而且v2ray本身不开tls也是能伪装host跟path的，应该不是只有tls能用吧。 我又看了下v2ray的tcp传输时开伪装应该是用http报文伪装，可能是用http的host请求头？但是伪装类型也可以是none，实在搞不清楚了。

我不明白你在想什么。如果你抓包不能找到host，那墙也找不到host。那就是很简单的逻辑咯，全部涂上马赛克，啥也看不到。那你问什么原理呢？我就告诉你，全部涂上马赛克，看不到了。就这么着，太简单了。

**Reply by tomandjerry27** [reply]
> > > > > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> > > > 
> > > > 
> > > > 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？
> > > 
> > > 
> > > 自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。 主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？
> > 
> > 
> > ws和tcp里都不放host的。host明文放在tls协议头中。path我不关心，毕竟host都明文了，path明不明根本不重要。 所以我一开始猜想你说的是ws-tls-tcp，没有tls的话，你只是能马赛克掉host，不能拿另一个host替换。那就变成思路一了。
> 
> 我查了tls不管是握手还是记录的报文头部都没有标放host的位置啊，ws跟host沾边的也只找到http报文里的host请求头。而且v2ray本身不开tls也是能伪装host跟path的，应该不是只有tls能用吧。 我又看了下v2ray的tcp传输时开伪装应该是用http报文伪装，可能是用http的host请求头？但是伪装类型也可以是none，实在搞不清楚了。

图片你看好没有?出于安全的考虑我打算删掉。

**Reply by XY-ShadowK** [reply]
> > > > > > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> > > > > 
> > > > > 
> > > > > 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？
> > > > 
> > > > 
> > > > 自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。 主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？
> > > 
> > > 
> > > ws和tcp里都不放host的。host明文放在tls协议头中。path我不关心，毕竟host都明文了，path明不明根本不重要。 所以我一开始猜想你说的是ws-tls-tcp，没有tls的话，你只是能马赛克掉host，不能拿另一个host替换。那就变成思路一了。
> > 
> > 
> > 我查了tls不管是握手还是记录的报文头部都没有标放host的位置啊，ws跟host沾边的也只找到http报文里的host请求头。而且v2ray本身不开tls也是能伪装host跟path的，应该不是只有tls能用吧。 我又看了下v2ray的tcp传输时开伪装应该是用http报文伪装，可能是用http的host请求头？但是伪装类型也可以是none，实在搞不清楚了。
> 
> 图片你看好没有?出于安全的考虑我打算删掉。

您删，我自己抓包看看

**Reply by tomandjerry27** [reply]
> > > > > > > 也就是说我是在报文里面放假的host跟path，那这样的话中间检测的时候不会用dns比对我的目的地址跟域名是否符合吗？还是说那样太耗资源没有那么干。
> > > > > > 
> > > > > > 
> > > > > > 你脑子转得真快，我都没想过这个情况。我觉得你说的都对，理论上说确实可以如此检测出假的host，但这么做也确实极耗资源。 但我要提醒你仔细看一下，我前面提到了“不论真假的域名”，所以你是可以去购买一个域名，然后使用真域名的。 另外，我觉得你太多想了。执剑人要把一个host加入黑名单是不需要理由的。你的真/假域名指向的服务器在境外，也没有备案信息（你拿他翻墙也不可能去备案吧），他们要把这个host从数据包里拎出来加入黑名单不需要再找理由了。你明白我的意思吗？
> > > > > 
> > > > > 
> > > > > 自己买域名的话国内云的大陆外主机建站都是免备案的，假域名可以用国内企业的外服域名（就是dns能发现异常）。 主要是ws只是长连接，报文跟http是一样的，但是tcp报文本身没有放host跟path的位置吧，还是说有常使用的约定俗成的放法呢？
> > > > 
> > > > 
> > > > ws和tcp里都不放host的。host明文放在tls协议头中。path我不关心，毕竟host都明文了，path明不明根本不重要。 所以我一开始猜想你说的是ws-tls-tcp，没有tls的话，你只是能马赛克掉host，不能拿另一个host替换。那就变成思路一了。
> > > 
> > > 
> > > 我查了tls不管是握手还是记录的报文头部都没有标放host的位置啊，ws跟host沾边的也只找到http报文里的host请求头。而且v2ray本身不开tls也是能伪装host跟path的，应该不是只有tls能用吧。 我又看了下v2ray的tcp传输时开伪装应该是用http报文伪装，可能是用http的host请求头？但是伪装类型也可以是none，实在搞不清楚了。
> > 
> > 
> > 图片你看好没有?出于安全的考虑我打算删掉。
> 
> 您删，我自己抓包看看

http里叫host，tls中叫sni（server name....）。host和sni名字不一样，其实是一回事。

**Comment by tomandjerry27** [comment]
我忘了说一个事情，这会儿我才想起来，虽然你根本也没问过。。。但应该你会想知道。
其实dns（我第一条回复中提到了dns）查询的时候就会暴露你真正要访问的域名了。

**Reply by XY-ShadowK** [reply]
dns本身是应用自己调的吧，socket是只绑地址端口的，访问的dns请求应该是浏览器自动调的系统api。

**Reply by XY-ShadowK** [reply]
> dns本身是应用自己调的吧，socket是只绑地址端口的，访问的dns请求应该是浏览器自动调的系统api。

而且同一个主机发那么多请求，可能同时一堆dns请求，也不大可能因为某个敏感dns后面结束后发了其他请求就判定有问题吧。更何况目的主机大概率不可能敏感，敏感直接访问不了换服了，跟客户端都没关系。
